apps/api/src/app/shared/services/in-memory-provider/in-memory-provider.service.ts

```diff
@@ -0,0 +1,73 @@
+import { Logger } from '@nestjs/common';
+
+import { getRedisInstance, getRedisProviderConfig, IRedisProviderConfig, Redis } from './redis-provider';
+
+import { ApiException } from '../../exceptions/api.exception';
+
+const LOG_CONTEXT = 'InMemoryCluster';
+
+export type InMemoryProviderClient = Redis;
+type InMemoryProviderConfig = IRedisProviderConfig;
+
+export class InMemoryProviderService {
+  public inMemoryProviderClient: InMemoryProviderClient;
+  public inMemoryProviderConfig: InMemoryProviderConfig;
+
+  constructor() {
+    if (!this.inMemoryProviderClient) {
+      this.inMemoryProviderClient = process.env.IN_MEMORY_CLUSTER_MODE
+        ? this.inMemoryClusterProviderSetup()
+        : this.inMemoryProviderSetup();
+    }
+  }
+
+  private inMemoryClusterProviderSetup(): InMemoryProviderClient {
+    Logger.verbose('In-memory cluster service set up', LOG_CONTEXT);
+
+    return {} as InMemoryProviderClient;
+  }
+
+  private inMemoryProviderSetup(): InMemoryProviderClient {
+    Logger.verbose('In-memory service set up', LOG_CONTEXT);
+
+    this.inMemoryProviderConfig = getRedisProviderConfig();
+    const { host, port, ttl } = getRedisProviderConfig();
+
+    if (!host) {
+      Logger.log('Missing host for in-memory provider', LOG_CONTEXT);
+      // TODO: Throw error
+    }
+
+    Logger.log(`Connecting to ${host}:${port}`, LOG_CONTEXT);
+
+    const inMemoryProviderClient = getRedisInstance();
+
+    inMemoryProviderClient.on('connect', () => {
+      Logger.log('REDIS CONNECTED', LOG_CONTEXT);
+    });
+
+    inMemoryProviderClient.on('error', (error) => {
+      Logger.error(error, LOG_CONTEXT);
+    });
+
+    return inMemoryProviderClient;
+  }
+
+  public async shutdown(): Promise<void> {
+    Logger.verbose('In-memory cluster service shutdown', LOG_CONTEXT);
+    await this.inMemoryProviderClient.quit();
+  }
+
+  /**
+   * This Nest.js hook allows us to execute logic on termination after signal.
+   * https://docs.nestjs.com/fundamentals/lifecycle-events#application-shutdown
+   *
+   * Enabled by:
+   *   app.enableShutdownHooks();
+   *
+   * in /apps/api/src/bootstrap.ts
+   */
+  public async onApplicationShutdown(signal): Promise<void> {
+    await this.shutdown();
+  }
```

Comment:
Very valid point! We need to add something similar to bullmq to gracefully finish processing jobs before shut down.